# **Exploring the imbalance among the Credit Card Fraud Dataset**

The dataset chosen is the well-known Credit Card Fraud , that consists in a set of anonymized credit card transations , that are either labeled as fradulent or genuine. From the Kaggle description, the dataset contains transactions made by credit cards in September 2013 by European cardholders. Moreover, this dataset presents transactions that occurred in two days, so that it has 492 frauds out of 284,807 transactions.

In [1]:
import kagglehub


path = kagglehub.dataset_download("mlg-ulb/creditcardfraud")

print("Path to dataset files:", path)

Path to dataset files: /home/andre/.cache/kagglehub/datasets/mlg-ulb/creditcardfraud/versions/3


In [2]:
import pandas as pd

df = pd.read_csv(path + "/creditcard.csv")
df.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0


In fact, the dataset contains only numerical input variables resulting from a PCA transformation. However, due to confidentiality issues, no background about the data was provided, as the features were overall encrypted. Features V1, V2, …, V28 are the principal components obtained through PCA. The only features that have not been transformed with PCA are 'Time' and 'Amount'.The 'Time' feature describes the seconds elapsed between each transaction and the first transaction in the dataset, while the feature 'Amount' is the transaction Amount, which can be used for example-dependant cost-sensitive learning. Feature 'Class' is the response variable and it takes value 1 in case of fraud and 0 otherwise.

## **Understanding the data**

As previosly stated, except for the 'transaction time' and 'amount', it is not known what are the meaning of the features in general (due to privacy reasons). Effectively, the Kaggle description of the data states that all the features, except for time and amount, went through a PCA transformation, that is a dimensionality reduction technique. It is important to take into account that , in order to implement a PCA transformation features need to be previously scaled. In this case, all the V features have been scaled (or, at least, it can be presumed that such operation was done).

In [4]:
df.shape

(284807, 31)

In [3]:
df.describe()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
count,284807.000000,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,...,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,284807.000000,284807.000000
mean,94813.859575,1.168375e-15,3.416908e-16,-1.379537e-15,2.074095e-15,9.604066e-16,1.487313e-15,-5.556467e-16,1.213481e-16,-2.406331e-15,...,1.654067e-16,-3.568593e-16,2.578648e-16,4.473266e-15,5.340915e-16,1.683437e-15,-3.660091e-16,-1.227390e-16,88.349619,0.001727
std,47488.145955,1.958696e+00,1.651309e+00,1.516255e+00,1.415869e+00,1.380247e+00,1.332271e+00,1.237094e+00,1.194353e+00,1.098632e+00,...,7.345240e-01,7.257016e-01,6.244603e-01,6.056471e-01,5.212781e-01,4.822270e-01,4.036325e-01,3.300833e-01,250.120109,0.041527
min,0.000000,-5.640751e+01,-7.271573e+01,-4.832559e+01,-5.683171e+00,-1.137433e+02,-2.616051e+01,-4.355724e+01,-7.321672e+01,-1.343407e+01,...,-3.483038e+01,-1.093314e+01,-4.480774e+01,-2.836627e+00,-1.029540e+01,-2.604551e+00,-2.256568e+01,-1.543008e+01,0.000000,0.000000
25%,54201.500000,-9.203734e-01,-5.985499e-01,-8.903648e-01,-8.486401e-01,-6.915971e-01,-7.682956e-01,-5.540759e-01,-2.086297e-01,-6.430976e-01,...,-2.283949e-01,-5.423504e-01,-1.618463e-01,-3.545861e-01,-3.171451e-01,-3.269839e-01,-7.083953e-02,-5.295979e-02,5.600000,0.000000
50%,84692.000000,1.810880e-02,6.548556e-02,1.798463e-01,-1.984653e-02,-5.433583e-02,-2.741871e-01,4.010308e-02,2.235804e-02,-5.142873e-02,...,-2.945017e-02,6.781943e-03,-1.119293e-02,4.097606e-02,1.659350e-02,-5.213911e-02,1.342146e-03,1.124383e-02,22.000000,0.000000
75%,139320.500000,1.315642e+00,8.037239e-01,1.027196e+00,7.433413e-01,6.119264e-01,3.985649e-01,5.704361e-01,3.273459e-01,5.971390e-01,...,1.863772e-01,5.285536e-01,1.476421e-01,4.395266e-01,3.507156e-01,2.409522e-01,9.104512e-02,7.827995e-02,77.165000,0.000000
max,172792.000000,2.454930e+00,2.205773e+01,9.382558e+00,1.687534e+01,3.480167e+01,7.330163e+01,1.205895e+02,2.000721e+01,1.559499e+01,...,2.720284e+01,1.050309e+01,2.252841e+01,4.584549e+00,7.519589e+00,3.517346e+00,3.161220e+01,3.384781e+01,25691.160000,1.000000


In [ ]:
df.isnull().sum().sum()

The dataset does not have any missing values , so , it doesn't need no imputation operations.

In [ ]:
df[df.duplicated()]

As it was noticed that the dataset has duplicates, these rows ended up being dropped, since they reffer to repeated transactions, which do not add value to the dataset analyis itself.

In [87]:
df.drop_duplicates(inplace=True)

Firstly, it can be plotted the distribution of the exposed features (that is , both time and Amount):

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

plt.figure(figsize=(12, 6))
sns.histplot(df['Time'], bins=50, kde=True, color='blue')
plt.title('Distribution of Time')
plt.xlabel('Time (seconds)')
plt.ylabel('Frequency')
plt.show()



Furthermore, for profilling purposes, it was inspected both the skewness and kurtosis values for both of the revealed features:

In [ ]:
df['Time'].skew()

In [ ]:
df['Time'].kurt()

In [ ]:

plt.figure(figsize=(10, 6))  
sns.boxplot(y=df['Amount'], color='lightblue')  
plt.title('Box Plot of Transaction Amount', fontsize=14)
plt.ylabel('Transaction Amount ($)', fontsize=12)
plt.xlabel('') 
plt.xticks(rotation=0)  
plt.yticks(rotation=0)  
plt.show()

In [ ]:
df['Amount'].skew()

In [ ]:
df['Amount'].kurt()

To perform a PCA, it is required to scale all of features that take part of such transformation. Despite of being unknown how the data was, the exposed features ( also the remaning ones to be scaled) should be scaled as well, as we are using later on a Logistic Regression Classifier, that requires the data to be scaled accordingly.

In [94]:

from sklearn.preprocessing import StandardScaler, RobustScaler

std_scaler = StandardScaler()
rob_scaler = RobustScaler()

df['Amount'] = rob_scaler.fit_transform(df['Amount'].values.reshape(-1,1))
df['Time'] = std_scaler.fit_transform(df['Time'].values.reshape(-1,1))


Next, as it was already expected that this dataset is , in fact, imbalanced, it is important to check the distribution of the target variable, that is, the 'Class' Variable:


In [ ]:
plt.figure(figsize=(6, 4))

sns.countplot(x='Class', data=df)
plt.title('Class Distributions \n (0: No Fraud || 1: Fraud)', fontsize=14)

class_counts = df['Class'].value_counts()
total = len(df)
percentages = (class_counts / total) * 100

for i, count in enumerate(class_counts):
    percentage = percentages[i]
    plt.text(i, count / 2, f'{percentage:.2f}%', ha='center', va='center', color='white', fontsize=12)

plt.show()

print("Class Distribution Percentages:")
for label, percentage in percentages.items():
    print(f"Class {label}: {percentage}%")
print("Imbalance Ratio: ", round(df['Class'].value_counts()[0]/ df['Class'].value_counts()[1], 1))

In order to track the datapoints of the minority class, and check their spread, a 2d-plot was done. In fact, such plot could point where are those data-points, which could indicate the most suitable sampling methods to be employed on this dataset. However, the data is high dimensional, which require the employ of a dimensional reduction technique. Firstly, it was considered a PCA (Principal Component Analysis), that is a linear method that projects high-dimensional data onto a lower-dimensional subspace by retaining the maximum variance in the data. It is fast, computationally efficient, and ideal for datasets with linear relationships between features, offering interpretable results. However, it struggles to capture nonlinear patterns or highlight clusters effectively (given that this is a binary classification problem, such struggle is not visible).


In [ ]:

import pandas as pd
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE

# Assuming your dataset is loaded in df
# Features and target
X = df.drop(columns=['Class'])
y = df['Class']

# Dimensionality Reduction
pca = PCA(n_components=2)
X_pca = pca.fit_transform(X)


In contrast, T-SNE (t-Distributed Stochastic Neighbor Embedding) is a nonlinear technique designed for exploring and visualizing complex datasets. As a matter of fact, t-SNE emphasizes preserving local structures by minimizing the divergence between probability distributions in high and low-dimensional spaces. This makes it excellent for revealing clusters and separability between classes, although it is computationally expensive and sensitive to hyperparameter choices, often losing global structure in the process.

In [97]:
tsne = TSNE(n_components=2, random_state=42)
X_tsne = tsne.fit_transform(X)


In [ ]:

plt.figure(figsize=(10, 7), dpi=100)
plt.scatter(X_pca[y == 0, 0], X_pca[y == 0, 1], label='Non-Fraud', alpha=0.6, s=10)
plt.scatter(X_pca[y == 1, 0], X_pca[y == 1, 1], label='Fraud', alpha=0.6, s=30, color='red')
plt.title('PCA Projection', fontsize=16)
plt.xlabel('Principal Component 1', fontsize=12)
plt.ylabel('Principal Component 2', fontsize=12)
plt.legend()
plt.tight_layout()
plt.show()


In [ ]:

plt.figure(figsize=(10, 7), dpi=100)
plt.scatter(X_tsne[y == 0, 0], X_tsne[y == 0, 1], label='Non-Fraud', alpha=0.6, s=10)
plt.scatter(X_tsne[y == 1, 0], X_tsne[y == 1, 1], label='Fraud', alpha=0.6, s=30, color='red')
plt.title('T-SNE Projection', fontsize=16)
plt.xlabel('T-SNE Component 1', fontsize=12)
plt.ylabel('T-SNE Component 2', fontsize=12)
plt.legend()
plt.tight_layout()
plt.show()


## Measuring Performance Metrics

### Splitting Data into Training and Test Sets

To test the performance of the classifiers over the dataset to further compare the impact of sampling methods on model performances, the dataset was split on a 70/30 proportion into a training set , and a test set. The split was stratified by the target variable to ensure that both the training and the test sets were following the same class proportions as the original dataset:

In [ ]:
from sklearn.model_selection import train_test_split


X_, y_ = df.drop(columns=['Class']), df['Class']
X_train, X_test, y_train, y_test = train_test_split(X_, y_, test_size=0.3,stratify=y_, random_state=42)


# Check class distribution in the train set
train_class_distribution = y_train.value_counts(normalize=True) * 100
print("Class distribution in the training set (%):")
print(train_class_distribution)

# Check class distribution in the test set
test_class_distribution = y_test.value_counts(normalize=True) * 100
print("\nClass distribution in the test set (%):")
print(test_class_distribution)

The class distributions shows us that, in both test and training set, the original proportions of the target variable's classes are maintained, given that these are close to the Class distributions both in the training set as in the test set. So, it is safe to say that the split was well done.
 

Next, there were chosen two well-known classifiers:  Logistic Regression and Decision Tree. In fact, it is known that these classifiers are well-suited for imbalanced datasets due to their respective strengths. Logistic Regression is configured to handle imbalanced data through threshold adjustments and regularization techniques, while Decision Tree classifiers excel at capturing non-linear relationships and focusing on the minority class using discriminative features. These classifiers are further tuned with the help of hyperparameter optimization technique, the GridSearch, to refine their performance.

In [101]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier


classifiers = {
    "LogReg": LogisticRegression(),
    "DecisionTree": DecisionTreeClassifier(),
    #"KNN": KNeighborsClassifier()
}


### **The meaning of the performance metrics**

In this context, each concept has a context-based meaning.

| **Term**            | **Definition**                                                                             |
|----------------------|-------------------------------------------------------------------------------------------|
| **True Positives**   | Correctly Classified Fraud Transactions                                                   |
| **False Positives**  | Incorrectly Classified Fraud Transactions                                                 |
| **True Negatives**   | Correctly Classified Non-Fraud Transactions                                               |
| **False Negatives**  | Incorrectly Classified Non-Fraud Transactions                                             |
| **Precision**        | True Positives / True Positives + False Positives            |            
| **Recall**           | True Positives / True Positives + False Negatives            |
| **Precision Meaning**| How precise is our model in detecting fraud transactions. That is, out of all the transactions the model flagged as fraud, the percentage of those who were actually fraud?                       |
| **Recall Meaning**   | The amount of fraud cases our model is able to detect. That is, out of all actual fraud cases, the percentage of those who were correctly identified  



#### **Precision-Recall Trade-off**

On the one hand, a model with a lower recall but an higher precision, is a model more selective. In fact, it is more likely to flag fewer transactions as fraud, which reduces false positives. However, it may result in missing some actual fraud cases (increasing false negatives). In a real-world perspective, it would be more suitable for scenarios where incorrectly flagging legitimate transactions has significant consequences, such as customer dissatisfaction or financial loss due to declined transactions.
On the other hand , a model with an higher recall , and a lower precision, is a model more aggressive. That means that the model will be flagging more transactions as fraud, which can increase the likelihood of catching all fraud cases, but it also may flag more legitimate transactions (increasing false positives) wrongly. In a real-world example, such model is more suitable for scenarios where detecting fraud is critical, and false positives are less problematic.

#### **The importance of ROC_AUC measure**

The ROC-AUC (Receiver Operating Characteristic - Area Under the Curve) is a crucial metric for evaluating binary classification models, particularly in tasks like fraud detection. It measures the model's ability to distinguish between classes (e.g., fraud vs. non-fraud) by assessing how well it ranks predictions based on their probabilities. Unlike threshold-dependent metrics, ROC-AUC evaluates performance across all thresholds, making it a comprehensive and fair measure. It balances the tradeoff between true positives and false positives, providing insights into the model's discriminative power. ROC-AUC is particularly valuable in imbalanced datasets, as it focuses on ranking predictions rather than absolute counts, making it less biased than metrics like accuracy. A high ROC-AUC value indicates strong model performance, while a relatively low value suggests the model performs no better than random guessing. 


### **Chossing the macro metric as the overall metric**

Macro metrics give equal importance to all classes by calculating metrics independently for each class and then taking their unweighted mean, regardless of class size. They are particularly useful when all classes are equally important or when dealing with imbalanced datasets. Macro metrics provide a clearer understanding of model performance across all classes, highlighting issues with minority class predictions, which is crucial in scenarios like fraud detection where misclassifying minority cases can have significant consequences.

In [102]:
from imblearn.pipeline import Pipeline as ImbPipeline
from imblearn.over_sampling import SMOTE
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer, precision_score, recall_score, f1_score, roc_auc_score, accuracy_score

def evaluate_baseline_classifiers(classifiers=classifiers, use_smote=False):
    scoring = {
        'accuracy': 'accuracy',
        'precision_macro': make_scorer(precision_score, average='macro'),
        'recall_macro': make_scorer(recall_score, average='macro'),
        'f1_macro': make_scorer(f1_score, average='macro'),
        'roc_auc': make_scorer(roc_auc_score)
    }

    for name, classifier in classifiers.items():
        print(f"\n{'='*40}")
        print(f"Evaluating: {name}")
        print(f"{'='*40}")
        
        if name == "LogReg":
            log_reg_params = {
                "classifier__penalty": ['l1', 'l2'],
                "classifier__C": [0.001, 0.01, 0.1, 1, 10, 100, 1000],
                "classifier__solver": ['liblinear']
            }

            if use_smote:
                pipeline = ImbPipeline([
                    ('smote', use_smote),
                    ('classifier', classifier)
                ])
            else:
                pipeline = ImbPipeline([
                    ('classifier', classifier)
                ])

            grid = GridSearchCV(pipeline, log_reg_params, scoring=scoring, refit='f1_macro', cv=5)
            grid.fit(X_train, y_train)
            best_model = grid.best_estimator_
            print(f"Best parameters for LogReg: {grid.best_params_}")
            print(f"Cross-validated F1 Macro Score: {grid.best_score_:.2f}")
            y_pred = best_model.predict(X_test)

        elif name == "DecisionTree":
            decision_tree_params = {
                "classifier__criterion": ["gini", "entropy"],
                "classifier__max_depth": [None, 5, 10, 20],
                "classifier__min_samples_split": [2, 5, 10],
                "classifier__min_samples_leaf": [1, 2, 4]
            }

            if use_smote:
                pipeline = ImbPipeline([
                    ('smote', use_smote),
                    ('classifier', classifier)
                ])
            else:
                pipeline = ImbPipeline([
                    ('classifier', classifier)
                ])

            grid = GridSearchCV(pipeline, decision_tree_params, scoring=scoring, refit='f1_macro', cv=5)
            grid.fit(X_train, y_train)
            best_model = grid.best_estimator_
            print(f"Best parameters for DecisionTree: {grid.best_params_}")
            print(f"Cross-validated F1 Macro Score: {grid.best_score_:.2f}")
            y_pred = best_model.predict(X_test)

        accuracy = accuracy_score(y_test, y_pred)
        precision = precision_score(y_test, y_pred, average='macro', zero_division=1)
        recall = recall_score(y_test, y_pred, average='macro')
        f1 = f1_score(y_test, y_pred, average='macro')
        roc_auc = roc_auc_score(y_test, y_pred)

        print("\nEvaluation metrics on the test set:")
        print(f"Accuracy: {accuracy:.2f}")
        print(f"Precision: {precision:.2f}")
        print(f"Recall: {recall:.2f}")
        print(f"F1 Score: {f1:.2f}")
        print(f"AUC-ROC: {roc_auc:.2f}")

In [103]:
evaluate_baseline_classifiers(classifiers)


## **Exploring Undersampling and Oversampling Approaches**

The [imbalanced-learn](https://imbalanced-learn.org/stable/index.html) toolbox offers a variety of implementations that corresponds to balancing techiques. In order to employ them in a more legible way, a set of wrapper functions were defined, each one corresponding to a sampling approach.

In [104]:
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import RandomOverSampler, SMOTE, ADASYN, SMOTEN, SVMSMOTE
from imblearn.combine import SMOTEENN, SMOTETomek

In [105]:
def random_undersampling(X, y):
    generator = RandomUnderSampler(random_state=42)
    X_res, y_res = generator.fit_resample(X, y)
    return X_res, y_res


def random_oversampling(X, y):
    generator = RandomOverSampler(random_state=42)
    X_res, y_res = generator.fit_resample(X, y)
    return X_res, y_res


def smote_oversampling(X, y):
    generator = SMOTE(random_state=42)
    X_res, y_res = generator.fit_resample(X, y)
    return X_res, y_res
def smote_n(X, y):
    generator = SMOTEN(random_state=42)
    X_res, y_res = generator.fit_resample(X, y)
    return X_res, y_res
def smote_enn(X, y):
    generator = SMOTEENN(random_state=42)
    X_res, y_res = generator.fit_resample(X, y)
    return X_res, y_res
def smotesvm(X, y):
    generator = SVMSMOTE(random_state=42)
    X_res, y_res = generator.fit_resample(X, y)
    return X_res, y_res
def smote_tomek(X, y):
    generator = SMOTETomek(random_state=42)
    X_res, y_res = generator.fit_resample(X, y)
    return X_res, y_res
def ADASYN_(X, y):
    generator = ADASYN(sampling_strategy='auto', random_state=42)
    X_res, y_res = generator.fit_resample(X, y)
    return X_res, y_res


### **Exploring Random Undersampling (RUS)**

In what follows, we will repeat the previous steps (fitting a classifier and evaluating performance) using different resampling techniques to tackle the class imbalance problem before model training. We start with **RandomUndersampling**, which randomly removes some of the majority class examples to achieve prevalence parity:

In [ ]:
from collections import Counter

X_train_rus, y_train_rus = random_undersampling(X_train, y_train)

print("Shape of X_train: ", X_train.shape)
print("Distribution of original training data: ", Counter(y_train))

print("Shape of X_train_rus: ", X_train_rus.shape)
print("Distribution of undersampled training data: ", Counter(y_train_rus))

For example, since the original training set consisted of **378 minority class examples** and **226980 majority class examples**, RUS randomly removes some majority class examples to keep only **378 majority examples** as well. The undersampled dataset now comprises 300 examples with balanced class priors.

In [107]:
evaluate_baseline_classifiers(classifiers, use_smote=RandomUnderSampler(random_state=42))

As already expected, given the disparity of the distribution of the target variable ( as previously seen before, 99.74% of the cases are classified as non-fraud), undersampling techniques will tend to erase the negative cases, witch results on a heavy information loss. In fact, the f1 scores of the class 1 reflects a under-performed model.

### **Exploring Random Oversampling (ROS)**

Rather than removing majority examples from the training, **Random Oversampling (ROS) inflates the minority class concepts** until the class priors are balanced. To that end, it randomly selects some minority class examples, creates exact replicas of these examples, and adds them to the training data.

In [ ]:
X_train_ros, y_train_ros = random_oversampling(X_train, y_train)

print("Shape of X_train: ", X_train.shape)
print("Distribution of original training data: ", Counter(y_train))

print("Shape of X_train_rus: ", X_train_ros.shape)
print("Distribution of undersampled training data: ", Counter(y_train_ros))

In [109]:
evaluate_baseline_classifiers(classifiers, use_smote=RandomOverSampler(random_state=42))

### **Exploring Synthetic Minority Oversampling Technique (SMOTE)**



Synthetic Minority Oversampling Technique (SMOTE) is a statistical technique for increasing the number of cases in your dataset in a balanced way. The component works by generating new instances from existing minority cases that you supply as input. This implementation of SMOTE does not change the number of majority cases.
The new instances are not just copies of existing minority cases. Instead, the algorithm takes samples of the feature space for each target class and its nearest neighbors. The algorithm then generates new examples that combine features of the target case with features of its neighbors. This approach increases the features available to each class and makes the samples more general.

In [ ]:

X_train_smote, y_train_smote = smote_oversampling(X_train, y_train)

print("Shape of X_train: ", X_train.shape)
print("Distribution of original training data: ", Counter(y_train))

print("Shape of X_train_smote: ", X_train_smote.shape)
print("Distribution of oversampled training data: ", Counter(y_train_smote))



In [111]:
evaluate_baseline_classifiers(classifiers, use_smote=SMOTE(sampling_strategy='minority',random_state=42))

#### **Exploring SMOTE Variants**

#### **Adaptive Synthetic Sampling (ADASYN)**

ADASYN works by generating synthetic samples for minority classes based on the feature space of the original dataset. It calculates the density distribution of each minority class sample and generates synthetic samples according to the density distribution. This adaptive approach ensures that more synthetic samples are generated for minority class samples that are harder to learn, thus improving the classification performance of machine learning models.

In [ ]:
X_train_smote_adasyn, y_train_smote_adasyn = ADASYN_(X_train, y_train)

print("Shape of X_train: ", X_train.shape)
print("Distribution of original training data: ", Counter(y_train))

print("Shape of X_train_smote_adasyn: ", X_train_smote_adasyn.shape)
print("Distribution of oversampled training data: ", Counter(y_train_smote_adasyn))



In [113]:
evaluate_baseline_classifiers(classifiers, use_smote=ADASYN(random_state=42))

#### **SMOTEN (Synthetic Minority Over-sampling Technique for Nominal and Continuous)** 

SMOTEN is an extension of the SMOTE algorithm designed to handle datasets with both categorical (nominal) and continuous features. Unlike traditional SMOTE, which works only with continuous data, SMOTEN generates synthetic minority class samples by preserving the integrity of categorical features while interpolating continuous features. This makes it an effective solution for balancing imbalanced datasets with mixed feature types, helping improve model performance by creating more realistic synthetic samples.

In [ ]:

X_train_smote_n, y_train_smote_n = smote_n(X_train, y_train)

print("Shape of X_train: ", X_train.shape)
print("Distribution of original training data: ", Counter(y_train))

print("Shape of X_train_smote_n: ", X_train_smote_n.shape)
print("Distribution of oversampled training data: ", Counter(y_train_smote_n))


In [115]:
evaluate_baseline_classifiers(classifiers, use_smote=SMOTEN(random_state=42))

#### **SMOTE SVM**

SMOTE SVM is a combination of the SMOTE (Synthetic Minority Over-sampling Technique) and Support Vector Machine (SVM) algorithms aimed at improving the performance of classification models on imbalanced datasets. SMOTE generates synthetic samples to balance the dataset, specifically for the minority class, while SVM is used as the classifier.
SMOTE is used to generate synthetic data points for the minority class, creating a more balanced dataset. Yet, SVM is then applied to this balanced dataset, where the model is trained to better distinguish between the minority and majority classes, leading to improved classification performance, especially for the minority class.
In summary, SMOTE SVM combines the strengths of synthetic over-sampling with SVM to address class imbalance and enhance the classifier's ability to detect minority class instances.

In [ ]:
X_train_smote_svm, y_train_smote_svm = smotesvm(X_train, y_train)

print("Shape of X_train: ", X_train.shape)
print("Distribution of original training data: ", Counter(y_train))

print("Shape of X_train_smote_svm: ", X_train_smote_svm.shape)
print("Distribution of oversampled training data: ", Counter(y_train_smote_svm))


In [ ]:
evaluate_baseline_classifiers(classifiers, use_smote=SVMSMOTE(random_state=42))

### **Combination of Over and under-sampling methods**

The imblearn.combine provides methods which combine over-sampling and under-sampling. In the imblearn.combine module, SMOTEEN and SMOTE Tomek are two advanced hybrid techniques that combine oversampling and undersampling to address imbalanced datasets effectively.

#### **SMOTEEN** 

This method starts with SMOTE for oversampling the minority class.
After oversampling, it uses the Edited Nearest Neighbors (ENN) method for undersampling. ENN removes samples whose class label differs from the majority of their nearest neighbors, effectively cleaning noisy or ambiguous samples.
SMOTEEN is particularly useful for addressing datasets with overlapping class distributions and significant noise, as it emphasizes dataset refinement after oversampling.

In [ ]:
X_train_smote_enn, y_train_smote_enn = smote_enn(X_train, y_train)

print("Shape of X_train: ", X_train.shape)
print("Distribution of original training data: ", Counter(y_train))

print("Shape of X_train_smote_enn: ", X_train_smote_enn.shape)
print("Distribution of oversampled training data: ", Counter(y_train_smote_enn))



In [ ]:
evaluate_baseline_classifiers(classifiers, use_smote=SMOTEENN(random_state=42))

#### **SMOTE-Tomek**

In a similar manner as SMOTEEN, this method first applies Synthetic Minority Oversampling Technique (SMOTE) to create synthetic samples for the minority class, thus increasing its representation in the dataset.
After oversampling, it applies the Tomek Links undersampling technique to clean the dataset. Tomek Links identify pairs of samples from different classes that are nearest neighbors. Removing such links eliminates borderline or noisy examples that might hinder the classifier's performance.
The combination of SMOTE and Tomek Links ensures the dataset is balanced while also reducing overlap between classes, leading to improved class separation.

In [ ]:
X_train_smote_tomek, y_train_smote_tomek = smote_tomek(X_train, y_train)

print("Shape of X_train: ", X_train.shape)
print("Distribution of original training data: ", Counter(y_train))

print("Shape of X_train_smote_tomek: ", X_train_smote_tomek.shape)
print("Distribution of oversampled training data: ", Counter(y_train_smote_tomek))



In [ ]:
evaluate_baseline_classifiers(classifiers, use_smote=SMOTETomek(random_state=42))

## **Results and Discussion**

As previously mentioned, Grid Search is employed here to systematically fine-tune the classifiers and ensure optimal performance. However, this process involves repeatedly training the same model with all possible combinations of the selected hyperparameters, which can be time-consuming. As a result, it was not feasible to run all evaluations in a single notebook. Instead, two different systems were used to obtain the results for each oversampling method.

Therefore, the performance of the models across all the employed oversampling methods is summarized in the following tables:



### Metrics Comparison Across Sampling Methods using a Logistic Regression Classifier

| **Sampling Method**  | **Accuracy** | **ROC-AUC** | **PRF (P/R/F1)** | **Best Parameters**                           |
|----------------------|--------------|-------------|------------------|-----------------------------------------------|
| Baseline             | 1.00         | 0.79        | 0.92/0.79/0.84   | {C: 10, penalty: l1, solver: liblinear}      |
| RUS                  | 0.98         | 0.93        | 0.53/0.93/0.55   | {C: 0.1, penalty: l1, solver: liblinear}     |
| ROS                  | 0.98         | 0.93        | 0.53/0.93/0.55   | {C: 0.001, penalty: l1, solver: liblinear}   |
| SMOTE                | 0.98         | 0.93        | 0.53/0.93/0.55   | {C: 0.001, penalty: l1, solver: liblinear}   |
| SMOTEN               | 1.00         | 0.81        | 0.93/0.81/0.86   | {C: 0.01, penalty: l2, solver: liblinear}    |
| SMOTEENN             | 0.98         | 0.93        | 0.53/0.93/0.55   | {C: 0.001, penalty: l1, solver: liblinear}   |
| SMOTESVM             | 0.99         | 0.92        | 0.58/0.92/0.63   | {C: 1, 'penalty: l1, classifier__solver: liblinear}  |
| SMOTE-Tomek          | 0.98         | 0.93        | 0.53/0.93/0.55   | {C: 0.001, penalty: l1, solver: liblinear} |
| ADASYN               | 0.92         | 0.91        | 0.51/0.91/0.50   | {C: 1000, penalty: l2, solver: liblinear}    |



### Metrics Comparison Across Sampling Methods using a Decision Tree Classifier

| **Sampling Method**  | **Accuracy** | **ROC-AUC** | **PRF (P/R/F1)** | **Best Parameters**                                                       |
|----------------------|--------------|-------------|------------------|---------------------------------------------------------------------------|
| Baseline             | 1.00         | 0.86        | 0.95/0.86/0.90   | criterion=entropy, max_depth=10, min_samples_leaf=4, min_samples_split=2 |
| RUS                  | 0.94         | 0.92        | 0.51/0.92/0.51   | criterion=gini, max_depth=5, min_samples_leaf=2, min_samples_split=5    |
| ROS                  | 1.00         | 0.86        | 0.89/0.86/0.87   | criterion=gini, max_depth=None, min_samples_leaf=2, min_samples_split=10 |
| SMOTE                | 1.00         | 0.87        | 0.75/0.87/0.80   | criterion=entropy, max_depth=None, min_samples_leaf=2, min_samples_split=5 |
| SMOTEN               | 1.00         | 0.85        | 0.96/0.85/0.90   | criterion=gini, max_depth=5, min_samples_leaf=1, min_samples_split=10    |
| SMOTEENN             | 1.00         | 0.88        | 0.75/0.88/0.80   | criterion=entropy, max_depth=None, min_samples_leaf=2, min_samples_split=2 |
| SMOTESVM             | 1.00         | 0.87        | 0.90/0.87/0.88   | criterion: 'entropy', max_depth: 20, min_samples_leaf: 4, min_samples_split: 10  |
| SMOTE-Tomek          | 1.00         | 0.87        | 0.75/0.87/0.80   | criterion=entropy, max_depth=None, min_samples_leaf=2, min_samples_split=10 |
| ADASYN               | 1.00         | 0.86        | 0.73/0.86/0.78   | criterion=entropy, max_depth=None, min_samples_leaf=2, min_samples_split=2 |


A comparison of the results reveals that the Decision Tree model generally performs better in terms of precision, recall, and F1-score for fraud detection, particularly in the baseline and SMOTEN configurations. The F1 score of 0.90 for fraud detection in the baseline model demonstrates the Decision Tree's strength in handling class imbalance. However, the ROC-AUC for Logistic Regression is also competitive, outperforming the Decision Tree in most oversampling methods (except for the baseline and SMOTEN).

On the one hand,  the Decision Tree maintains a good balance between performance and interpretability. Thus, considering that the Decision Tree classifier performed better overall, especially in fraud detection, it is a more effective model for credit card fraud detection. This is particularly true when applying the SMOTEN sampling technique, which yields the best precision across all cases. In a real-world scenario, this precision would be crucial for avoiding the false labeling of legitimate transactions as fraud. As mentioned before, falsely flagging legitimate transactions as fraudulent can have significant negative consequences, such as customer dissatisfaction or financial losses from declined legitimate transactions. However, it is also important to recognize that some fraudulent transactions may go unnoticed (false negatives), which could also be problematic.

On the other hand, Logistic Regression consistently demonstrates better recall scores, especially when using oversampling methods. Although it does not maintain a balance with precision (with precision being notably low in most oversampling cases compared to the baseline), this classifier would be more suitable in scenarios where identifying as many fraudulent transactions as possible is critical. In such situations, the consequences of false positives are relatively less severe.

## **Conclusions**


The work conducted not only aimed to explore and address the class imbalance issue present in the chosen dataset, but also focused on identifying the most appropriate model, alongside its optimal hyperparameters, for effectively classifying transactions as fraudulent or non-fraudulent while maintaining reasonable model complexity. Additionally, considerable attention was given to selecting the most suitable sampling techniques to handle the imbalance in the data. Given that the dataset mirrors a real-world problem, various practical scenarios were considered throughout the process.

Through a careful examination, multiple models, each paired with their optimal hyperparameters, were identified and tested across different oversampling methods. These models demonstrated their suitability for different scenarios, effectively addressing the precision-recall trade-off that is particularly critical in fraud detection tasks. By optimizing both the models and the sampling methods, the performance of each classifier was maximized in ways that allowed the system to strike a balance between precision and recall, ensuring reliable fraud detection while minimizing false positives.

Ultimately, the work confirms that a careful combination of suitable models, hyperparameters, and sampling methods can effectively balance the precision-recall trade-off, enabling the development of a robust fraud detection system that performs well in real-world applications. This approach not only improves model performance but also ensures that it is adaptable to varying levels of fraud detection requirements and potential consequences.

## References

- To consolidate the concepts discussed herein, please refer to the following:

    - Kaggle. 2018. [Credit Card Fraud Detection](https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud)
    - Jason Brownlee. (n.d.). ROC Curves and Precision-Recall Curves for Imbalanced Classification.
    - GeeksforGeeks. (2024). How do decision trees work with unbalanced datasets?
    - The imbalanced-learn contributors. (n.d.). SMOTE — Version 0.13.0.
    - Janiobachmann. (2019). Credit Fraud || Dealing with Imbalanced Datasets
    - Scikit-learn contributors. (n.d.). DecisionTreeClassifier. 
    - Scikit-learn contributors. (n.d.). GridSearchCV.
    - Scikit-learn. (n.d.). LogisticRegression.
    - Y. Sun, A. K. Wong, and M. S. Kamel. (2007). "Logistic Regression for Class Imbalanced Datasets." IEEE Transactions on Systems, Man, and Cybernetics, 37(4), 914–922. DOI: 10.1109/TSMCA.2006.885801